src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/ExecutableScenarioComposerImpl.java

```diff
@@ -0,0 +1,90 @@
+package executor.service.service.parallelflowexecutor.impls.subscribers;
+
+import executor.service.model.proxy.ProxyConfigHolder;
+import executor.service.model.scenario.Scenario;
+import executor.service.model.service.ExecutableScenario;
+import executor.service.queue.ProxyQueue;
+import executor.service.queue.ScenarioQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ExecutableScenarioComposerImpl implements ExecutableScenarioComposer {
+
+    private static final Logger log = LoggerFactory.getLogger(ExecutableScenarioComposerImpl.class);
+
+    private final ProxyQueue proxyQueue;
+    private final ScenarioQueue scenarioQueue;
+    private final ProxyConfigHolder defaultProxy;
+
+    public ExecutableScenarioComposerImpl(ProxyQueue proxyQueue, ScenarioQueue scenarioQueue, ProxyConfigHolder defaultProxy) {
+        this.proxyQueue = proxyQueue;
+        this.scenarioQueue = scenarioQueue;
+        this.defaultProxy = defaultProxy;
+    }
+
+    @Override
+    public List<ExecutableScenario> composeExecutableScenarios() {
+        try {
+            List<ProxyConfigHolder> proxies = proxyQueue.getAllProxy();
+            List<Scenario> scenarios = scenarioQueue.getAllScenario();
+            if (!scenarios.isEmpty() && !proxies.isEmpty()) {
+                return composeNotEmptyList(proxies, scenarios);
+            } else if (!scenarios.isEmpty()) {
+                return composeByDefaultProxy(defaultProxy, scenarios);
+            } else {
+                proxies.forEach(proxyQueue::addProxy);
```

Comment:
Changing the logic here will improve performance.

Reply:
![image](https://github.com/GFLCourses5/worker/assets/140727960/38017804-23fe-464e-9cac-90d056e525ff)

This is a bad practice (38 line), you first take data from the collection and then return it, this can be avoided by checking the collection by size.

Reply:
Do you remember that we work with different queues in parallel threads?  I don't know queues is full or empty, in next moment size of queue can changed and this algorithm wouldn't work directly. Though the queues are synchronized, but we don't block them on read/write additional mechanisms.